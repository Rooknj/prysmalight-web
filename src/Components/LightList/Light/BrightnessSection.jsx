import React from "react";
import PropTypes from "prop-types";
import Slider from "./SmoothSlider";
import styled from "styled-components";
import BrightnessIcon from "@material-ui/icons/BrightnessLow";

const IconDiv = styled.div`
  margin-top: auto;
  margin-bottom: auto;
  padding-right: 1rem;
  padding-left: 1rem;
`;

const StyledDiv = styled.div`
  margin-bottom: 1rem;
  margin-left: 1rem;
  margin-right: 1rem;
  display: flex;
`;

const StyledSlider = styled(Slider)`
  padding-top: 1rem;
  padding-bottom: 1rem;
`;

const propTypes = {
  connected: PropTypes.number,
  brightness: PropTypes.number,
  onBrightnessChange: PropTypes.func
};

const defaultProps = {};

const BrightnessSection = ({ brightness, onBrightnessChange, connected }) => {
  return (
    <React.Fragment>
      <StyledDiv onClick={e => e.stopPropagation()}>
        <IconDiv>
          <BrightnessIcon color="primary" />
        </IconDiv>
        <StyledSlider
          value={brightness}
          min={0}
          max={100}
          step={1}
          onChange={onBrightnessChange}
          disabled={connected !== 2}
        />
      </StyledDiv>
    </React.Fragment>
  );
};

BrightnessSection.propTypes = propTypes;
BrightnessSection.defaultProps = defaultProps;

export default BrightnessSection;
