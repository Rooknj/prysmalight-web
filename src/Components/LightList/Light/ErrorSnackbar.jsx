import React from "react";
import Snackbar from "@material-ui/core/Snackbar";
import IconButton from "@material-ui/core/IconButton";
import CloseIcon from "@material-ui/icons/Close";
import styled from "styled-components";
import Typography from "@material-ui/core/Typography";

const StyledIconButton = styled(IconButton)`
  padding: ${({ theme }) => theme.spacing.unit / 4}px;
`;

const propTypes = {};
const defaultProps = {};

class ErrorSnackbar extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      open: true
    };
  }
  handleClose = (event, reason) => {
    if (reason === "clickaway") {
      return;
    }
    this.setState({ open: false });
  };
  render() {
    const { message } = this.props;
    return (
      <React.Fragment>
        <Snackbar
          anchorOrigin={{
            vertical: "bottom",
            horizontal: "left"
          }}
          open={this.state.open}
          autoHideDuration={6000}
          onClose={this.handleClose}
          ContentProps={{
            "aria-describedby": "message-id"
          }}
          message={
            <span id="message-id">
              <Typography variant="body1" color="error">
                {message}
              </Typography>
            </span>
          }
          action={[
            <StyledIconButton
              key="close"
              aria-label="Close"
              color="inherit"
              onClick={this.handleClose}
            >
              <CloseIcon />
            </StyledIconButton>
          ]}
        />
      </React.Fragment>
    );
  }
}
ErrorSnackbar.propTypes = propTypes;
ErrorSnackbar.defaultProps = defaultProps;
export default ErrorSnackbar;
