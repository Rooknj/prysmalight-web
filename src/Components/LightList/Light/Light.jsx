import React from "react";
import PropTypes from "prop-types";

import Card from "@material-ui/core/Card";
import LightHeader from "./LightHeader/LightHeader";
import BrightnessSlider from "./BrightnessSlider/BrightnessSlider";
import Collapse from "@material-ui/core/Collapse";

import LightDialog from "./LightDialog/LightDialog";

import styled from "styled-components";

const StyledCardWrapper = styled.div`
  min-width: 18rem;
  margin: 0 auto;
  padding: 0.5rem;
`;

const propTypes = {
  light: PropTypes.shape({
    id: PropTypes.string.isRequired,
    connected: PropTypes.number,
    state: PropTypes.string,
    brightness: PropTypes.number,
    color: PropTypes.shape({
      r: PropTypes.number.isRequired,
      g: PropTypes.number.isRequired,
      b: PropTypes.number.isRequired
    }),
    effect: PropTypes.string,
    speed: PropTypes.number,
    supportedEffects: PropTypes.array
  }),
  loading: PropTypes.bool,
  onStateChange: PropTypes.func.isRequired,
  onBrightnessChange: PropTypes.func.isRequired,
  onColorChange: PropTypes.func.isRequired,
  onEffectChange: PropTypes.func.isRequired
};

const defaultProps = {
  light: {
    id: "",
    connected: 0,
    state: "OFF",
    brightness: 0,
    color: {
      r: 0,
      g: 0,
      b: 0
    }
  }
};

class Light extends React.Component {
  state = {
    open: false
  };

  handleOpen = () => {
    this.setState({ open: true });
  };

  handleClose = () => {
    this.setState({ open: false });
  };

  render() {
    const { light, loading, onStateChange, onBrightnessChange } = this.props;
    return (
      <React.Fragment>
        <LightDialog
          open={this.state.open}
          onClose={this.handleClose}
          {...this.props}
        />
        <StyledCardWrapper>
          <Card onClick={this.handleOpen}>
            <LightHeader
              id={light.id}
              color={light.color}
              connected={light.connected}
              state={light.state}
              onChange={onStateChange}
              waiting={loading}
            />
            <Collapse
              in={light.state === "ON" ? true : false}
              timeout="auto"
              unmountOnExit
            >
              <BrightnessSlider
                connected={light.connected}
                brightness={light.brightness}
                onBrightnessChange={onBrightnessChange}
              />
            </Collapse>
          </Card>
        </StyledCardWrapper>
      </React.Fragment>
    );
  }
}

Light.propTypes = propTypes;
Light.defaultProps = defaultProps;

export default Light;
