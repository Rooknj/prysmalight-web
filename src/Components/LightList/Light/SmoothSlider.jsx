import React from "react";
import Slider from "@material-ui/lab/Slider";

export default class SmoothSlider extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      value: props.value
    };
  }

  handleChange = (_, value) => {
    this.setState({ value });
    this.props.onChange(_, value);
  };

  componentDidUpdate(prevProps) {
    if (this.props.value !== prevProps.value) {
      this.setState({ value: this.props.value });
    }
  }

  render() {
    const { value, onChange, ...props } = this.props;
    return (
      <Slider
        value={this.state.value}
        onChange={this.handleChange}
        {...props}
      />
    );
  }
}
