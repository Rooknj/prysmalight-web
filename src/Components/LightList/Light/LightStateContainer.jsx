import React from "react";
import PropTypes from "prop-types";

import { throttle } from "lodash";
import Light from "./Light";

// Need to set the throttle speed to 0 when running tests so that setLight gets called on every test
let throttleSpeed;
if (process.env.REACT_APP_ENV === "test") {
  throttleSpeed = 0;
} else {
  throttleSpeed = 500;
}

const throttleSetLight = throttle((setLight, newLight, oldLight) => {
  // newColor will keep the __typename from the old light while assigning new color values
  const newColor = Object.assign({}, oldLight.color, newLight.color);
  // Set up the optimistic response
  const optimisticResponse = {
    __typename: "Mutation",
    setLight: {
      __typename: "Light",
      ...oldLight,
      ...newLight,
      color: newColor
    }
  };

  const lightId = newLight.id;
  const lightData = Object.assign({}, newLight);
  delete lightData["id"];
  setLight({
    variables: {
      lightId,
      lightData
    },
    optimisticResponse
  });
}, throttleSpeed);

class LightStateContainer extends React.Component {
  handleStateChange = e => {
    const { setLight, light } = this.props;
    const newLight = {
      id: light.id,
      state: e.target.checked ? "ON" : "OFF"
    };
    throttleSetLight(setLight, newLight, light);
  };

  handleBrightnessChange = (_, brightness) => {
    const { setLight, light } = this.props;
    const newLight = {
      id: light.id,
      brightness
    };
    throttleSetLight(setLight, newLight, light);
  };

  handleColorChange = ({ rgb: { r, g, b } }) => {
    const { setLight, light } = this.props;
    if (r === light.color.r && g === light.color.g && b === light.color.b) {
      return;
    }
    const newLight = {
      id: light.id,
      color: { r, g, b }
    };
    throttleSetLight(setLight, newLight, light);
  };

  handleEffectChange = e => {
    const { setLight, light } = this.props;
    const newLight = {
      id: light.id,
      [e.target.name]: e.target.value
    };
    throttleSetLight(setLight, newLight, light);
  };

  render() {
    const { light, loading } = this.props;
    return (
      <Light
        light={light}
        loading={loading}
        onStateChange={this.handleStateChange}
        onBrightnessChange={this.handleBrightnessChange}
        onColorChange={this.handleColorChange}
        onEffectChange={this.handleEffectChange}
      />
    );
  }
}

LightStateContainer.propTypes = {
  light: PropTypes.shape({
    id: PropTypes.string.isRequired,
    connected: PropTypes.number,
    state: PropTypes.string,
    brightness: PropTypes.number,
    color: PropTypes.shape({
      r: PropTypes.number.isRequired,
      g: PropTypes.number.isRequired,
      b: PropTypes.number.isRequired
    }),
    effect: PropTypes.string,
    speed: PropTypes.number,
    supportedEffects: PropTypes.array
  }).isRequired,
  loading: PropTypes.bool
};

export default LightStateContainer;
