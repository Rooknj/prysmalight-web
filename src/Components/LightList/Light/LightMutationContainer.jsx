import React from "react";

import { Mutation } from "react-apollo";
import { SET_LIGHT } from "../../graphqlConstants";
import LightStateContainer from "./LightStateContainer";
import ErrorSnackbar from "./ErrorSnackbar";

class LightMutationContainer extends React.Component {
  render() {
    // The mutation will automatically update the cache
    return (
      <Mutation mutation={SET_LIGHT}>
        {(mutate, { data, loading, error }) => {
          return (
            <React.Fragment>
              <LightStateContainer
                {...this.props}
                setLight={mutate}
                lightData={data}
                loading={loading}
                error={error}
              />
              {error && (
                <ErrorSnackbar
                  message={`Error changing ${this.props.light.id}`}
                />
              )}
            </React.Fragment>
          );
        }}
      </Mutation>
    );
  }
}

export default LightMutationContainer;
