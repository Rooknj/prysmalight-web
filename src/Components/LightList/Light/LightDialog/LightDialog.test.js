import React from "react";
import { shallow, mount } from "enzyme";
import Light from "./LightDialog";
import { createSerializer } from "enzyme-to-json";
import { ThemeProvider } from "styled-components";

// Add serializer to allow enzyme to create snapshots
expect.addSnapshotSerializer(createSerializer({ mode: "deep" }));

const mockTheme = { spacing: { unit: 2 } };

// TODO: Test rendering with actual mock light data
it("renders without crashing", () => {
  shallow(
    <Light
      onStateChange={() => true}
      onBrightnessChange={() => true}
      onColorChange={() => true}
      onEffectChange={() => true}
    />
  );
});

it("matches shallow snapshot", () => {
  const shallowed = shallow(
    <Light
      onStateChange={() => true}
      onBrightnessChange={() => true}
      onColorChange={() => true}
      onEffectChange={() => true}
    />
  );
  expect(shallowed).toMatchSnapshot();
});

it("matches mounted snapshot", () => {
  const mounted = mount(
    <ThemeProvider theme={mockTheme}>
      <Light
        onStateChange={() => true}
        onBrightnessChange={() => true}
        onColorChange={() => true}
        onEffectChange={() => true}
      />
    </ThemeProvider>
  );
  expect(mounted).toMatchSnapshot();
});
