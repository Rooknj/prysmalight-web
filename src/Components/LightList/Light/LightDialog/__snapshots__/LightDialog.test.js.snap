// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches mounted snapshot 1`] = `""`;

exports[`matches shallow snapshot 1`] = `
<ZoomDialog
  fullScreen={true}
  open={false}
>
  <LightHeader
    light={
      Object {
        "brightness": 0,
        "color": Object {
          "b": 0,
          "g": 0,
          "r": 0,
        },
        "connected": 0,
        "id": "",
        "state": "OFF",
      }
    }
    onBrightnessChange={[Function]}
    onMore={[Function]}
    onStateChange={[Function]}
  />
  <Light
    light={
      Object {
        "brightness": 0,
        "color": Object {
          "b": 0,
          "g": 0,
          "r": 0,
        },
        "connected": 0,
        "id": "",
        "state": "OFF",
      }
    }
    onBrightnessChange={[Function]}
    onColorChange={[Function]}
    onEffectChange={[Function]}
    onSpeedChange={[Function]}
    onStateChange={[Function]}
    open={false}
  />
  <OptionsDrawerContainer
    light={
      Object {
        "brightness": 0,
        "color": Object {
          "b": 0,
          "g": 0,
          "r": 0,
        },
        "connected": 0,
        "id": "",
        "state": "OFF",
      }
    }
    onClose={[Function]}
    onOpen={[Function]}
    onRenameSelected={[Function]}
    open={false}
  />
  <Mutation
    mutation={
      Object {
        "definitions": Array [
          Object {
            "directives": Array [],
            "kind": "OperationDefinition",
            "name": Object {
              "kind": "Name",
              "value": "setLight",
            },
            "operation": "mutation",
            "selectionSet": Object {
              "kind": "SelectionSet",
              "selections": Array [
                Object {
                  "alias": undefined,
                  "arguments": Array [
                    Object {
                      "kind": "Argument",
                      "name": Object {
                        "kind": "Name",
                        "value": "lightId",
                      },
                      "value": Object {
                        "kind": "Variable",
                        "name": Object {
                          "kind": "Name",
                          "value": "lightId",
                        },
                      },
                    },
                    Object {
                      "kind": "Argument",
                      "name": Object {
                        "kind": "Name",
                        "value": "lightData",
                      },
                      "value": Object {
                        "kind": "Variable",
                        "name": Object {
                          "kind": "Name",
                          "value": "lightData",
                        },
                      },
                    },
                  ],
                  "directives": Array [],
                  "kind": "Field",
                  "name": Object {
                    "kind": "Name",
                    "value": "setLight",
                  },
                  "selectionSet": Object {
                    "kind": "SelectionSet",
                    "selections": Array [
                      Object {
                        "directives": Array [],
                        "kind": "FragmentSpread",
                        "name": Object {
                          "kind": "Name",
                          "value": "lightFields",
                        },
                      },
                    ],
                  },
                },
              ],
            },
            "variableDefinitions": Array [
              Object {
                "defaultValue": undefined,
                "directives": Array [],
                "kind": "VariableDefinition",
                "type": Object {
                  "kind": "NonNullType",
                  "type": Object {
                    "kind": "NamedType",
                    "name": Object {
                      "kind": "Name",
                      "value": "String",
                    },
                  },
                },
                "variable": Object {
                  "kind": "Variable",
                  "name": Object {
                    "kind": "Name",
                    "value": "lightId",
                  },
                },
              },
              Object {
                "defaultValue": undefined,
                "directives": Array [],
                "kind": "VariableDefinition",
                "type": Object {
                  "kind": "NonNullType",
                  "type": Object {
                    "kind": "NamedType",
                    "name": Object {
                      "kind": "Name",
                      "value": "LightInput",
                    },
                  },
                },
                "variable": Object {
                  "kind": "Variable",
                  "name": Object {
                    "kind": "Name",
                    "value": "lightData",
                  },
                },
              },
            ],
          },
          Object {
            "directives": Array [],
            "kind": "FragmentDefinition",
            "name": Object {
              "kind": "Name",
              "value": "lightFields",
            },
            "selectionSet": Object {
              "kind": "SelectionSet",
              "selections": Array [
                Object {
                  "alias": undefined,
                  "arguments": Array [],
                  "directives": Array [],
                  "kind": "Field",
                  "name": Object {
                    "kind": "Name",
                    "value": "id",
                  },
                  "selectionSet": undefined,
                },
                Object {
                  "alias": undefined,
                  "arguments": Array [],
                  "directives": Array [],
                  "kind": "Field",
                  "name": Object {
                    "kind": "Name",
                    "value": "name",
                  },
                  "selectionSet": undefined,
                },
                Object {
                  "alias": undefined,
                  "arguments": Array [],
                  "directives": Array [],
                  "kind": "Field",
                  "name": Object {
                    "kind": "Name",
                    "value": "connected",
                  },
                  "selectionSet": undefined,
                },
                Object {
                  "alias": undefined,
                  "arguments": Array [],
                  "directives": Array [],
                  "kind": "Field",
                  "name": Object {
                    "kind": "Name",
                    "value": "state",
                  },
                  "selectionSet": undefined,
                },
                Object {
                  "alias": undefined,
                  "arguments": Array [],
                  "directives": Array [],
                  "kind": "Field",
                  "name": Object {
                    "kind": "Name",
                    "value": "brightness",
                  },
                  "selectionSet": undefined,
                },
                Object {
                  "alias": undefined,
                  "arguments": Array [],
                  "directives": Array [],
                  "kind": "Field",
                  "name": Object {
                    "kind": "Name",
                    "value": "color",
                  },
                  "selectionSet": Object {
                    "kind": "SelectionSet",
                    "selections": Array [
                      Object {
                        "alias": undefined,
                        "arguments": Array [],
                        "directives": Array [],
                        "kind": "Field",
                        "name": Object {
                          "kind": "Name",
                          "value": "r",
                        },
                        "selectionSet": undefined,
                      },
                      Object {
                        "alias": undefined,
                        "arguments": Array [],
                        "directives": Array [],
                        "kind": "Field",
                        "name": Object {
                          "kind": "Name",
                          "value": "g",
                        },
                        "selectionSet": undefined,
                      },
                      Object {
                        "alias": undefined,
                        "arguments": Array [],
                        "directives": Array [],
                        "kind": "Field",
                        "name": Object {
                          "kind": "Name",
                          "value": "b",
                        },
                        "selectionSet": undefined,
                      },
                    ],
                  },
                },
                Object {
                  "alias": undefined,
                  "arguments": Array [],
                  "directives": Array [],
                  "kind": "Field",
                  "name": Object {
                    "kind": "Name",
                    "value": "effect",
                  },
                  "selectionSet": undefined,
                },
                Object {
                  "alias": undefined,
                  "arguments": Array [],
                  "directives": Array [],
                  "kind": "Field",
                  "name": Object {
                    "kind": "Name",
                    "value": "speed",
                  },
                  "selectionSet": undefined,
                },
                Object {
                  "alias": undefined,
                  "arguments": Array [],
                  "directives": Array [],
                  "kind": "Field",
                  "name": Object {
                    "kind": "Name",
                    "value": "supportedEffects",
                  },
                  "selectionSet": undefined,
                },
              ],
            },
            "typeCondition": Object {
              "kind": "NamedType",
              "name": Object {
                "kind": "Name",
                "value": "Light",
              },
            },
          },
        ],
        "kind": "Document",
        "loc": Object {
          "end": 336,
          "start": 0,
        },
      }
    }
    onCompleted={[Function]}
    onError={[Function]}
  >
    <Component />
  </Mutation>
</ZoomDialog>
`;
